src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * \file array/cpu/labor_pick.h
+ * \brief Template implementation for layerwise pick operators.
+ */
+
+#ifndef DGL_ARRAY_CPU_LABOR_PICK_H_
+#define DGL_ARRAY_CPU_LABOR_PICK_H_
+
+#include <dgl/random.h>
+#include <dgl/array.h>
+#include <dmlc/omp.h>
+#include <dgl/runtime/parallel_for.h>
+#include <parallel_hashmap/phmap.h>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <numeric>
+#include <memory>
+#include <utility>
+#include <cmath>
+#include <iostream>
+#include <pcg_random.hpp>
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+// Template for picking non-zero values row-wise.
+template <typename IdxType, typename FloatType>
+std::pair<COOMatrix, FloatArray> CSRLaborPick(
+    CSRMatrix mat,
+    IdArray NIDs,
+    IdArray rows,
+    int64_t __num_picks,
+    FloatArray prob,
+    IdArray random_seed,
+    IdArray cnt,
+    int importance_sampling) {
+  using namespace aten;
+  const IdxType* indptr = static_cast<IdxType*>(mat.indptr->data);
+  const IdxType* indices = static_cast<IdxType*>(mat.indices->data);
+  const IdxType* data = CSRHasData(mat)? static_cast<IdxType*>(mat.data->data) : nullptr;
+  const IdxType* rows_data = static_cast<IdxType*>(rows->data);
+  const IdxType* nids = NIDs->shape[0] ? static_cast<IdxType*>(NIDs->data) : nullptr;
+  const auto num_rows = rows->shape[0];
+  const IdxType num_picks = __num_picks;
+  const auto& ctx = mat.indptr->ctx;
+
+  constexpr bool linear_c_convergence = false;
+
+  const bool weights = !IsNullArray(prob);
+  if (importance_sampling >= 0)
+    importance_sampling += weights;
+  auto A_arr = prob;
+  FloatType *A = static_cast<FloatType*>(A_arr->data);
+  constexpr double eps = 0.0001;
+  constexpr FloatType ONE = 1;
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* cs = static_cast<FloatType*>(cs_array->data);
+  FloatArray ds_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* ds = static_cast<FloatType*>(ds_array->data);
+
+  IdxType max_d = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
+    max_d = std::max(act_d, max_d);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_d;
```

Comment:
Do a sanity check on the weight values (`d > 0`)?

Reply:
If d == 0, then accumulate is run with an empty range, then it will return 0.

Reply:
But in L94 you are dividing `num_picks` by `d`.  Is that OK though?

Reply:
d is a double so it is.

Reply:
If d is zero, then that value is not used at all.

Reply:
I have run the code will different datasets and it didn't crash even once over hundreds of training runs.

Reply:
I see.